src/logging/checkpoint_manager.cpp

```diff
@@ -0,0 +1,450 @@
+    CopySerializeInput record_decode((const void *) buffer.get(), length);
+
+    size_t persist_epoch_id = (size_t) record_decode.ReadLong();
+
+
+    int database_count = record_decode.ReadInt();
+
+    database_structures.resize(database_count);
+
+    for (oid_t database_idx = 0; database_idx < (oid_t)database_count; ++database_idx) {
+      
+      int table_count = record_decode.ReadInt();
+
+      for (oid_t table_idx = 0; table_idx < (oid_t)table_count; table_idx++) {
+        int tile_group_count = record_decode.ReadInt();
+
+        database_structures[database_idx].push_back(tile_group_count);
+      }
+    }
+
+    // Safely close the file
+    bool res = LoggingUtil::CloseFile(file_handle);
+    if (res == false) {
+      LOG_ERROR("Cannot close pepoch file");
+      exit(EXIT_FAILURE);
+    }
+
+    return persist_epoch_id;
+  }
+
+  void CheckpointManager::RecoverCheckpoint(const size_t &epoch_id, const std::vector<std::vector<size_t>> &database_structures) {
+
+    // open files
+    FileHandle ***file_handles = new FileHandle**[database_structures.size()];
+    
+    for (oid_t database_idx = 0; database_idx < database_structures.size(); database_idx++) {
+    
+      file_handles[database_idx] = new FileHandle*[database_structures[database_idx].size()];
+
+      for (oid_t table_idx = 0; table_idx < database_structures[database_idx].size(); table_idx++) {
+        
+        file_handles[database_idx][table_idx] = new FileHandle[max_checkpointer_count_];
+
+        for (size_t virtual_checkpointer_id = 0; virtual_checkpointer_id < max_checkpointer_count_; virtual_checkpointer_id++) {
+          
+          std::string file_name = GetCheckpointFileFullPath(virtual_checkpointer_id % checkpointer_count_, virtual_checkpointer_id, database_idx, table_idx, epoch_id);
+         
+          bool success =
+              LoggingUtil::OpenFile(file_name.c_str(), "rb", file_handles[database_idx][table_idx][virtual_checkpointer_id]);
+          
+          PL_ASSERT(success == true);
+          if (success != true) {
+            LOG_ERROR("create file failed!");
+            exit(-1);
+          }
+        }
+      }
+    }
+
+    // before parallel recovery, we may need to prepare tables.
+    // PrepareTables(database_structures);
+
+    /////////////////////////////////////////////////////////////////////
+    // after obtaining the database structures, we can start recovering checkpoints in parallel.
+    recovery_pools_.resize(recovery_thread_count_);
+    for (size_t i = 0; i < recovery_thread_count_; ++i) {
+      recovery_pools_[i].reset(new type::EphemeralPool()); //BACKEND_TYPE_MM
+    }
+    
+    std::vector<std::unique_ptr<std::thread>> recovery_threads;
+    recovery_threads.resize(recovery_thread_count_);
+    for (size_t i = 0; i < recovery_thread_count_; ++i) {
+      recovery_threads[i].reset(new std::thread(&CheckpointManager::RecoverCheckpointThread, this, i, epoch_id, std::ref(database_structures), file_handles));
```

Comment:
Should we really be spawning threads here? Is there a central place where we could be asking for threads?